crates/ika-types/src/messages_dwallet_mpc.rs

```diff
@@ -159,13 +192,99 @@ pub struct DWalletMPCMessage {
 /// Used to make sure no message is being processed twice.
 #[derive(Clone, Debug, Serialize, Deserialize, Hash, PartialEq, Eq, Ord, PartialOrd)]
 pub struct DWalletMPCMessageKey {
+    /// The serialized message.
+    pub message: MPCMessageSlice,
     /// The authority (Validator) that sent the message.
     pub authority: AuthorityName,
     pub session_id: ObjectID,
     /// The MPC round number, starts from 0.
     pub round_number: usize,
 }
 
+/// Collects and reconstructs `MPCMessage`s from all parties across multiple rounds.
+/// This struct is useful for aggregating fragmented MPC messages from different
+/// parties, round by round, and reassembling them once all parts are received.
+#[derive(Clone)]
+pub struct MPCSessionMessagesCollector {
+    /// Each index corresponds to a round. Each round maps `PartyID` to its message builder.
+    pub messages: Vec<HashMap<PartyID, MPCMessageBuilder>>,
+}
+
+impl MPCSessionMessagesCollector {
+    /// Creates a new, empty message collector.
+    pub fn new() -> Self {
+        Self {
+            messages: Vec::new(),
+        }
+    }
+
+    /// Adds a message from a given party to the collector.
+    ///
+    /// If a complete message is reconstructed, it is returned as `Some(message)`.
+    /// Otherwise, returns `None`.
+    /// If the round number is beyond the current vector length, a new entry is created.
+    pub fn add_message(
+        &mut self,
+        party_id: PartyID,
+        message: DWalletMPCMessage,
+    ) -> Option<Vec<u8>> {
+        let messages_len = self.messages.len();
+        let round_number = message.round_number;
+        let sequence_number = message.message.sequence_number;
+        let message_slice = message.message.clone();
+
+        match self.messages.get_mut(round_number) {
+            Some(party_to_msg) => {
+                let entry = party_to_msg.entry(party_id).or_insert_with(|| {
+                    let mut builder = MPCMessageBuilder {
+                        messages: MessageState::Incomplete(
+                            vec![(sequence_number, message_slice.clone())]
+                                .into_iter()
+                                .collect(),
+                        ),
+                    };
+                    builder.build_message();
+                    builder
+                });
+
+                entry.add_message(message_slice.clone());
+                entry.build_message();
```

Comment:
Its a weird design. No use for these being two different functions as they will always be called together